src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,146 @@
 package core.basesyntax;
 
 import java.util.List;
+import java.util.Objects;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size;
+    private Node<T> head;
+    private Node<T> tail;
+
+    public static class Node<T> {
+        private Node<T> prev;
+        private T value;
+        private Node<T> next;
+
+        public Node(Node<T> prev, T value, Node<T> next) {
+            this.prev = prev;
+            this.value = value;
+            this.next = next;
+        }
+    }
+
     @Override
     public void add(T value) {
+        if (size == 0) {
+            head = new Node<>(null, value, null);
+            tail = head;
+        } else {
+            Node<T> newNode = new Node<>(tail, value, null);
+            tail.next = newNode;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexForAdd(index);
+        if (index == size) {
+            add(value);
+        } else if (index == 0) {
+            Node<T> newNode = new Node<>(null, value, getNode(index));
+            head = newNode;
+            newNode.next.prev = newNode;
+            size++;
+        } else {
+            Node<T> newNode = new Node<>(getNode(index).prev, value, getNode(index));
+            newNode.prev.next = newNode;
+            newNode.next.prev = newNode;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexForGet(index);
+        return getNode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexForGet(index);
+        T oldNodeValue = get(index);
+        getNode(index).value = value;
+        return oldNodeValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexForGet(index);
+        T oldNodeValue = getNode(index).value;
+        if (size == 1) {
+            head = null;
+            tail = null;
+        } else {
+            if (index == 0) {
+                head = head.next;
+                head.prev = null;
+            } else if (index == (size - 1)) {
+                tail = tail.prev;
+                tail.next = null;
+            } else {
+                getNode(index).prev.next = getNode(index).next;
+                getNode(index).prev = getNode(index).prev.prev;
+            }
```

Comment:
could we simplify this? And I think we should use `getNode(index)` method only ones here (It's pretty heavy method, if we will use it several time without extreme necessity it will effect on performance)